### PR TITLE
GEOMESA-2841 ILIKE queries with invalid regexp throw tserver errors

### PR DESCRIPTION
**Pull request by jnh5y**
Signed-off-by: Jim Hughes <jnh5y@ccri.com>

### CODE REVIEW

File: geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala [hunk 1]
```diff
@@ -316,6 +317,17 @@ class QueryPlanFilterVisitor(sft: SimpleFeatureType) extends DuplicatingFilterVi
     }
   }
 
+  override def visit(filter: PropertyIsLike, extraData: Any): AnyRef = {
+    val pattern = filter.getLiteral
+    try {
+      Pattern.compile(pattern)
+    } catch {
+      case e: PatternSyntaxException =>
+        throw new IllegalArgumentException("The regex for the (i)like filter is invalid.", e)
```
Comment (elahrvivaz) [inline geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala:326]: can we add the pattern into the exception message?

File: geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala [hunk 2]
```diff
@@ -316,6 +317,17 @@ class QueryPlanFilterVisitor(sft: SimpleFeatureType) extends DuplicatingFilterVi
     }
   }
 
+  override def visit(filter: PropertyIsLike, extraData: Any): AnyRef = {
+    val pattern = filter.getLiteral
+    try {
+      Pattern.compile(pattern)
```
Comment (elahrvivaz) [inline geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala:323]: i think we might need to use the LikeToRegexConverter to handle the %'s and all that appropriately: https://github.com/geotools/geotools/blob/master/modules/library/main/src/main/java/org/geotools/filter/LikeFilterImpl.java#L166
https://github.com/geotools/geotools/blob/e8cf256a9e7ad034863d08a2bc680371f2b774b9/modules/library/main/src/main/java/org/geotools/filter/LikeToRegexConverter.java#L30

File: geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala [hunk 3]
```diff
@@ -316,6 +318,16 @@ class QueryPlanFilterVisitor(sft: SimpleFeatureType) extends DuplicatingFilterVi
     }
   }
 
+  override def visit(filter: PropertyIsLike, extraData: Any): AnyRef = {
+    try {
+      new LikeToRegexConverter(filter).getPattern
+    } catch {
+      case e: Exception =>
+        throw new IllegalArgumentException(s"The regex filter (${ECQL.toCQL(filter)}) for the (i)like filter is invalid.", e)
```
Comment (jahhulbert) [inline geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala:326]: is there any possibility that ECQL.toCQL() might throw an exception here as well?

Comment (elahrvivaz) [reply geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala:326]: you should be able to use `filter.getLiteral` to return the raw string value, which would be safe

File: geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala [hunk 4]
```diff
@@ -316,6 +318,16 @@ class QueryPlanFilterVisitor(sft: SimpleFeatureType) extends DuplicatingFilterVi
     }
   }
 
+  override def visit(filter: PropertyIsLike, extraData: Any): AnyRef = {
+    try {
+      new LikeToRegexConverter(filter).getPattern
```
Comment (elahrvivaz) [inline geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala:323]: do you still need to pass it into `Pattern.compile`? looking at the LikeToRegexConverter I don't think it throws any validation errors, but I could be wrong.

Comment (jnh5y) [reply geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala:323]: Good question; doing this was sufficient to throw an exception in the provided unit test.  I think that exception may have been in the GeoTools code which was trying to parse / extract the regex?

Comment (jnh5y) [reply geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala:323]: I'll toss in the call to Pattern.compile.

File: geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala (hunk 3 above)
Comment (jnh5y) [reply geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/visitor/QueryPlanFilterVisitor.scala:326]: Ok, switched to filter.getLiteral.

**Comment by elahrvivaz** [comment]
merged as 68cedbf402f902d9086178d831baa604ab2d7648

**Pull request closed without merging**